### PR TITLE
Add Datasource interface and RPC implementation

### PR DESCRIPTION
**Pull request by sylviamoss**
This adds the new plugin type Datasource.

_merge after https://github.com/hashicorp/packer/pull/10440 is rebased with master [done ✔️ ]_

### CODE REVIEW

**Review by azr — APPROVED** [review]
This sure looks good to me 👍🏼  nice one !

**Comment by sylviamoss** [comment]
Don't merge yet! I'm writing tests for https://github.com/hashicorp/packer/pull/10440 and updating this one.

**Review by nywilken — APPROVED** [review]
This looks good to me. Nicely done with the testing. It would be great to sit with you and walk-through the RPC communication bits, for my learning :smile:

File: packer/maps.go [hunk 1]
```diff
@@ -72,3 +72,26 @@ func (mob MapOfBuilder) List() []string {
 	}
 	return res
 }
+
+type MapOfDatasource map[string]func() (Datasource, error)
+
+func (mob MapOfDatasource) Has(dataSource string) bool {
```
[the remark this answers is not in the document]
Comment (sylviamoss) [reply packer/maps.go]: good catch! thanks

**Pull request merged**